### PR TITLE
Allow partial vault withdrawal

### PR DESCRIPTION
**Pull request by yusufidimaina9989**

### CODE REVIEW

**Comment by xhliu** [comment]
New tx diagram here
https://miro.com/app/board/uXjVKmBUtlY=/

File: src/contracts/vaultCompleteWithdrawal.ts [hunk 1]
```diff
@@ -33,49 +26,65 @@ export class VaultCompleteWithdrawal extends SmartContract {
...
         const prevTxId = hash256(
             prevTxVer +
-            prevTxInputContract +
-            prevTxInputFee +
-            toByteString('02') + vaultAmt + vaultSPK + dust + targetSPK +
-            prevTxLocktime
+                prevTxInputContract +
+                prevTxInputFee +
+                toByteString('02') +
+                withdrawalAmt +
```
Comment (msinkec) [inline src/contracts/vaultCompleteWithdrawal.ts]: Prev tx is a withdrwal trigger tx. This tx doesn't yet have payout output.

File: src/contracts/vaultCompleteWithdrawal.ts [hunk 2]
```diff
@@ -33,49 +26,65 @@ export class VaultCompleteWithdrawal extends SmartContract {
...
-            feePrevout
+            prevTxId + toByteString('00000000') + feePrevout
         )
         assert(hashPrevouts == shPreimage.hashPrevouts, 'hashPrevouts mismatch')
 
-        // Enforce outputs
+        // Enforce outputs.
         const hashOutputs = sha256(
-            vaultAmt + targetSPK
+            withdrawalAmt +
```
Comment (msinkec) [inline src/contracts/vaultCompleteWithdrawal.ts]: Make vault output first and payout second. Looks nicer in diagrams.

File: src/contracts/vaultCompleteWithdrawal.ts (hunk 1 above)
Comment (msinkec) [reply src/contracts/vaultCompleteWithdrawal.ts]: Also, if remaining amount is 0, then the vault gets destroyed, e.g. only the payout output is enforced.

File: src/contracts/vaultTriggerWithdrawal.ts [hunk 3]
```diff
@@ -22,9 +22,7 @@ export class VaultTriggerWithdrawal extends SmartContract {
      * @param withdrawalPubKey - Public key, used for withdrawal.
      *
      */
-    constructor(
-        withdrawalPubKey: PubKey,
-    ) {
+    constructor(withdrawalPubKey: PubKey) {
```
Comment (msinkec) [inline src/contracts/vaultTriggerWithdrawal.ts]: Trigger contract shouldn't do payout. For now, I think this contract doesn't need any change at all.

File: tests/vault.test.ts [hunk 4]
```diff
@@ -191,7 +191,7 @@ describe('Test SmartContract `Vault`', () => {
         let interpreter = new btc.Script.Interpreter()
         let flags = btc.Script.Interpreter.SCRIPT_VERIFY_WITNESS | btc.Script.Interpreter.SCRIPT_VERIFY_TAPROOT
         let res = interpreter.verify(new btc.Script(''), tx0.outputs[0].script, tx1, 0, flags, witnesses, tx0.outputs[0].satoshis)
-        expect(res).to.be.true
+        expect(res).not.throw
```
Comment (msinkec) [inline tests/vault.test.ts]: res needs to be true.

**Review by msinkec — CHANGES_REQUESTED** [review]

**Pull request merged**